docs/source/python/memory.rst

```diff
@@ -277,6 +277,97 @@ types than with normal Python file objects.
    !rm example.dat
    !rm example2.dat
 
+Efficiently Writing and Reading Arrow Arrays
+--------------------------------------------
+
+Being optimized for zero copy and memory mapped data, Arrow allows to easily
+read and write arrays consuming the minimum amount of resident memory.
+
+When writing and reading raw Arrow data, we can use the Arrow File Format
+or the Arrow Streaming Format.
+
+To dump an array to file, you can use the :meth:`~pyarrow.ipc.new_file`
+which will provide a new :class:`~pyarrow.ipc.RecordBatchFileWriter` instance
+that can be used to write batches of data to that file.
+
+For example to write an array of 100M integers, we could write it in 1000 chunks
+of 100000 entries:
+
+.. ipython:: python
+
+    BATCH_SIZE = 100000
+    NUM_BATCHES = 1000
+
+    schema = pa.schema([pa.field('nums', pa.int32())])
+
+    with pa.OSFile('bigfile.arrow', 'wb') as sink:
+        with pa.ipc.new_file(sink, schema) as writer:
+            for row in range(NUM_BATCHES):
+                batch = pa.record_batch([pa.array(range(BATCH_SIZE), type=pa.int32())], schema)
+                writer.write(batch)
+
+record batches support multiple columns, so in practice we always write the
+equivalent of a :class:`~pyarrow.Table`.
+
+Writing in batches is effective because we in theory need to keep in memory only
+the current batch we are writing. But when reading back, we can be even more effective
+by directly mapping the data from disk and avoid allocating any new memory on read.
+
+Under normal conditions, reading back our file will consume a few hundred megabytes
+of memory:
+
+.. ipython:: python
+
+    with pa.OSFile('bigfile.arrow', 'rb') as source:
+        loaded_array = pa.ipc.open_file(source).read_all()
+
+    print("LEN:", len(loaded_array))
+    print("RSS: {}MB".format(pa.total_allocated_bytes() >> 20))
+
+To more efficiently read big data from disk, we can memory map the file, so that
+the arrow can directly reference the data mapped from disk and avoid having to
+allocate its own memory.
+In such case the operating system will be able to page in the mapped memory
+lazily and page it out without any write back cost when under pressure,
+allowing to more easily read arrays bigger than the total memory.
+
+.. ipython:: python
+
+    with pa.memory_map('bigfile.arrow', 'r') as source:
+        loaded_array = pa.ipc.open_file(source).read_all()
+    print("LEN:", len(loaded_array))
+    print("RSS: {}MB".format(pa.total_allocated_bytes() >> 20))
+
+Equally we can write back to disk the loaded array without consuming any
+extra memory thanks to the fact that iterating over the array will just
+scan through the data without the need to make copies of it
```

Comment:
While this is true, I'm not sure the use case of writing back identical memory-mapped data is very interesting to talk about. Usually, you would write back data after some amount of processing (e.g. filtering).

Reply:
I agree it would be nice to show a more "interesting" example where memory-mapping can be useful.

It's also not really clear to me (as a memory-mapping noob) what the above exactly means if you do some operation on the memory mapped data (that is not just writing it back unmodified).

Reply:
If you want an example with a clear performance benefit then a partial IPC read is a pretty good one.  The IPC reader today does not column selection into I/O filtering.  In other words, even if you read only a few columns it will still "read" the entire file.  Since it doesn't access the memory for the undesired columns you can see a benefit in memory mapping.

One could conceivably implement a smarter IPC reader that does selectively read I/O (with prebuffering) and probably close the gap somewhat although the overhead of figuring out and issuing all the small reads may still leave an advantage to the memory mapped method.

Reply:
I did some experiments here https://github.com/apache/arrow/issues/10138 which demonstrate the effect.

Reply:
@jorisvandenbossche the memory mapping only provides benefits if you don't alter data. If you applied any transformation to the data, the data wouldn't be equal to the one on disk anymore and thus you would lose all benefits of memory mapping because the kernel would no longer be able to use the memory mapped file instead of the swap file when in need to page out. I guess we can remove the "write back" section of the documentation if you think it doesn't provide much value.

My primary goal was mostly to say "if you need to open a big ipc format file, open it using memory mapping or you will just face a OOM" the writing back section was to reinforce the concept but doesn't really add any additional value. I'm mostly interested in shipping an addition to the docs that documents that concept somewhere and I don't want to make perfect the enemy of good enough so I'm ok with deferring to other PRs any part that doesn't get obvious consensus.

Reply:
As mentioned in https://github.com/apache/arrow/pull/10266#discussion_r673995490 I got rid of the "write back" section which seemed to just open a can of worms without providing much practical value for a reader as it was.